### ISSUE TITLE
NullPointerException while fetching tables

### ISSUE THREAD

**Issue by gschlager** [jOOQ/jOOQ#4313]
There's a `NullPointerException` when `regexFlags` in `AbstractDatabase` is not set.

```
ERROR org.jooq.util.AbstractDatabase - Error while fetching tables
java.lang.NullPointerException: null
	at java.util.ArrayList.<init>(ArrayList.java:177) ~[na:1.8.0_45]
	at org.jooq.util.AbstractDatabase.pattern(AbstractDatabase.java:243) ~[jooq-meta-3.6.1.jar:na]
	at org.jooq.util.AbstractDatabase.filterExcludeInclude(AbstractDatabase.java:1010) ~[jooq-meta-3.6.1.jar:na]
	at org.jooq.util.AbstractDatabase.filterExcludeInclude(AbstractDatabase.java:985) ~[jooq-meta-3.6.1.jar:na]
	at org.jooq.util.AbstractDatabase.getTables(AbstractDatabase.java:693) ~[jooq-meta-3.6.1.jar:na]
```

**Comment by lukaseder** [comment]
Thanks for reporting. Interesting, how can that be `null` ("not set")? By default, the value is initialised to an empty list:

```java
    public List<RegexFlag> getRegexFlags() {
        if (regexFlags == null) {
            regexFlags = new ArrayList<RegexFlag>();
        }
        return this.regexFlags;
    }
```

How did you set up your code generator?

**Comment by gschlager** [comment]
I'm using jooq-meta for extracting database metadata, but I'm not using the code generator and its XML based configuration. I'm creating a database with `JooqDatabaseFactory.createInstance()` and didn't have to initialize the `regexFlags` up until now... (I just upgraded to 3.6)

**Comment by lukaseder** [comment]
Hmm, interesting. Where is that `JooqDatabaseFactory` from? A third party library?

**Comment by gschlager** [comment]
Yes, that's part of a tool we developed at work and now that I see it I should probably change that part of the code since #3782 was fixed a long time ago.

**Comment by lukaseder** [comment]
Aha, I see - so you're not going through the `GenerationTool`'s initialisation, then? That would explain it. We should take proper measures inside of `org.jooq.util.AbstractDatabase` to prevent this from happening.

**Comment by lukaseder** [comment]
... I guess the same is true for the `includes` and `excludes` arrays?

**Comment by gschlager** [comment]
Yes, I believe so. We had to set a bunch of properties in order to get it to work.
```java
database.setIncludes(new String[] { ".*" });
database.setExcludes(new String[] {});
database.setRecordVersionFields(new String[] {});
database.setRecordTimestampFields(new String[] {});
database.setConfiguredCustomTypes(new ArrayList<CustomType>());
database.setConfiguredEnumTypes(new ArrayList<EnumType>());
database.setConfiguredForcedTypes(new ArrayList<ForcedType>());
database.setRegexFlags(new ArrayList<RegexFlag>());
```

I'm not sure if all of those are still required, but that's what it looks right now.

**Comment by lukaseder** [comment]
OK, I see. You shouldn't need to set those values, of course. The fix will lazy-initialise all arrays and lists to empty objects as it should be.

**Comment by lukaseder** [comment]
The fix will be merged to 3.6.2 (#4317), 3.5.5 (#4318), and 3.4.6 (#4319)
